microsoft/testsuites/kselftest/kselftest.py

```diff
@@ -0,0 +1,274 @@
+import os
+import re
+from dataclasses import dataclass
+from pathlib import PurePath, PurePosixPath
+from typing import Any, Dict, List
+
+from assertpy import assert_that
+
+from lisa import Environment, notifier
+from lisa.base_tools.uname import Uname
+from lisa.executable import Tool
+from lisa.messages import SubTestMessage, TestStatus, create_test_result_message
+from lisa.node import Node
+from lisa.operating_system import CBLMariner
+from lisa.testsuite import TestResult
+from lisa.tools import Cp, Git, Ls, Make, RemoteCopy, Tar
+from lisa.tools.whoami import Whoami
+from lisa.util import LisaException, SkippedException, find_groups_in_lines
+
+
+@dataclass
+class KselftestResult:
+    name: str = ""
+    status: TestStatus = TestStatus.QUEUED
+    exit_value: int = 0
+
+
+class Kselftest(Tool):
+    _MARINER_KERNEL_SRC_REPO = "https://github.com/microsoft/CBL-Mariner-Linux-Kernel"
+    _KERNEL_REPO_NAME = "CBL-Mariner-Linux-Kernel"
+    _KSELFTEST_TAR_PATH = (
+        "/build/kselftest/kselftest_install/kselftest-packages/kselftest.tar.gz"
+    )
+
+    # kselftest result log has "ok" and "not ok" prefixes, use regex to filter them
+    # [{'status': 'not ok', 'name': 'step_after_suspend_test', 'info': '# exit=126'}]
+    _RESULT_KSELFTEST_OK_REGEX = re.compile(
+        r"^(?P<status>(not ok|ok))\s+\d+\s+selftests:\s+\S+:\s+(?P<name>\S+)\s*(?P<info>.*)$"  # noqa: E501
+    )
+
+    @property
+    def command(self) -> str:
+        return str(self._command)
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def _check_exists(self) -> bool:
+        return self.node.tools[Ls].path_exists(str(self._remote_tar_path), sudo=True)
+
+    def __init__(
+        self, node: Node, kselftest_file_path: str, *args: Any, **kwargs: Any
+    ) -> None:
+        super().__init__(node, *args, **kwargs)
+
+        # tar file path specified in yml
+        self._tar_file_path = kselftest_file_path
+        if self._tar_file_path:
+            self._remote_tar_path = self.get_tool_path(
+                use_global=True
+            ) / os.path.basename(self._tar_file_path)
+        else:
+            self._remote_tar_path = (
+                self.get_tool_path(use_global=True)
+                / f"{self._KERNEL_REPO_NAME}{self._KSELFTEST_TAR_PATH}"
+            )
+
+        # command to run kselftests
+        if not self._tar_file_path:
+            self._command = self.get_tool_path(use_global=True) / "run_kselftest.sh"
+        else:
+            self._command = (
+                self.get_tool_path(use_global=True)
+                / "kselftest-packages/run_kselftest.sh"
+            )
+
+    # install common dependencies
+    def _install(self) -> bool:
+        if not isinstance(self.node.os, CBLMariner):
+            raise SkippedException("kselftests are supported on Mariner VMs only.")
```

Comment:
1. `kselftests test case is supported Mariner`. We need to claim it's a limitation of LISA, instead of kselftests.
2. Here should raise unsupported exception, and convert it to skippped exception in test case level. The tools are not related to test purpose, so it shouldn't raise any skipped exception.

Reply:
This is a limitation of kselftests. Our focus for now is to run kselftests automatically on Mariner VMs. The main use case of this PR is for MarinerMOS team and LSG kernel team to validate kernels automatically in our pipelines which uses mariner build container. This feature can be extended for different distros in the future.

Initially we thought we'd just skip running tests if a non-mariner distro is used. Do you recommend changing this to UnsupportedDistroException instead?

Reply:
> 1. `kselftests test case is supported Mariner`. We need to claim it's a limitation of LISA, instead of kselftests.
> 2. Here should raise unsupported exception, and convert it to skippped exception in test case level. The tools are not related to test purpose, so it shouldn't raise any skipped exception.

I have fixed all the comments, please re-review. Thanks!

Reply:
I mean not to use `SkippedException` in tool level. The tool is to execute, it shouldn't contain test purpose. The test case decide a case to fail or skip.